v3/mathjax3-ts/util/LinkedList.ts

```diff
@@ -0,0 +1,356 @@
+/*************************************************************
+ *
+ *  Copyright (c) 2017 The MathJax Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * @fileoverview  Implement a generic LinkedList object.
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+/*****************************************************************/
+/*
+ *  A symbol used to mark the special node used to indicate
+ *  the start and end of the list.
+ */
+const END = Symbol();
+
+/*
+ * Shorthand type for the functions used to sort the data items
+ */
+export type SortFn<DataClass> = (a: DataClass, b: DataClass) => boolean;
+
+/*****************************************************************/
+/*
+ *  The ListItem interface (for a specific type of data item)
+ *
+ *  These are the items in the doubly-linked list.
+ */
+
+export class ListItem<DataClass> {
+    /*
+     * The data for the list item
+     */
+    public data: DataClass | symbol;
+
+    /*
+     * Pointers to the next and previous items in the list
+     */
+    public next: ListItem<DataClass> = null;
+    public prev: ListItem<DataClass> = null;
+
+    /*
+     * @param{any} data  The data to be stored in the list item
+     * @constructor
+     */
+    constructor (data: any = null) {
+        this.data = data;
+    }
+}
+
+/*****************************************************************/
+/*
+ *  The ListList class interface
+ */
+
+export interface LinkedListClass<DataClass> {
+    new(...args: DataClass[]): LinkedList<DataClass>;
+}
+
+/*****************************************************************/
+/*
+ *  Implements the generic ListList class
+ */
+
+export class LinkedList<DataClass> {
+
+    /*
+     * The linked list
+     */
+    protected list: ListItem<DataClass>;
```

Comment:
What is wrong with `list`?  this points to the actual linked list.  I could see using `root`, though `dummy` doesn't make sense to me, and I don't see that `anchor` describes what it actually is.